### PR TITLE
update group_rows_by, GroupedMatrixTable

### PR DESCRIPTION
**Pull request by catoverdrive**
@tpoterba I still need to add some tests for GroupedMatrixTable, but what do you think about this?

### CODE REVIEW

File: python/hail/matrixtable.py [hunk 1]
```diff
@@ -1663,20 +1663,24 @@ def group_rows_by(self, *exprs, **named_exprs):
 
         Notes
         -----
-        The `key_expr` argument can either be a string referring to a row-indexed field
-        of the dataset, or an expression that will become the new row key.
+        All complex expressions (expressions that aren't simply row fields) must
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: indicate that nested fields are OK too

File: python/hail/matrixtable.py [hunk 2]
```diff
@@ -1663,20 +1663,24 @@ def group_rows_by(self, *exprs, **named_exprs):
...
+        be passed in as a keyword argument and are assigned to a field.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: "must be passed as named expressions" (that's a concept we use elsewhere I think)

File: python/hail/matrixtable.py [hunk 3]
```diff
@@ -1687,13 +1691,20 @@ def group_rows_by(self, *exprs, **named_exprs):
...
+                raise ExpressionException("method 'group_rows_by' found duplicate field: {}".format(key))
+            new_keys.append(key)
+
+        ds = self.annotate_rows(**named_exprs)
+        for key in named_exprs.keys():
+            if key in new_keys:
+                raise ExpressionException("method 'group_rows_by' found duplicate field: {}".format(key))
+            new_keys.append(key)
+
+
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: aggressive whitespace

File: python/hail/matrixtable.py [hunk 4]
```diff
@@ -1687,13 +1691,20 @@ def group_rows_by(self, *exprs, **named_exprs):
...
-            groups.append((key, e))
-        for k, e in named_exprs.items():
-            e = to_expr(e)
-            analyze('MatrixTable.group_rows_by', e, self._row_indices)
-            groups.append((k, e))
 
-        return GroupedMatrixTable(self, groups, self._row_indices)
+            if key in new_keys:
+                raise ExpressionException("method 'group_rows_by' found duplicate field: {}".format(key))
+            new_keys.append(key)
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: I think this case is handled wrong -- suppose I pass ds.foo.bar.gene. what happens? Let's make sure that case is tested.

File: python/hail/matrixtable.py [hunk 5]
```diff
@@ -1713,20 +1724,22 @@ def group_cols_by(self, *exprs, **named_exprs):
 
         Notes
         -----
-        The `key_expr` argument can either be a string referring to a column-indexed
-        field of the dataset, or an expression that will become the new column key.
+        All complex expressions (expressions that aren't simply column fields) must
+        be passed in as a keyword argument and are assigned to a field.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: same comments

File: python/hail/matrixtable.py [hunk 6]
```diff
@@ -1713,20 +1724,22 @@ def group_cols_by(self, *exprs, **named_exprs):
...
 
         Parameters
         ----------
-        key_expr : str or :class:`.Expression`
-            Field name or expression to use as new column key.
+        exprs : args of :obj:`str` or :class:`.Expression`
+            Column fields to group on.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: s/on/by

File: python/hail/matrixtable.py [hunk 7]
```diff
@@ -1713,20 +1724,22 @@ def group_cols_by(self, *exprs, **named_exprs):
...
+        named_exprs : varargs of :class:`.Expression`
+            Column-indexed expressions to group on.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: same

File: python/hail/matrixtable.py [hunk 8]
```diff
@@ -1663,20 +1663,24 @@ def group_rows_by(self, *exprs, **named_exprs):
...
-        of the dataset, or an expression that will become the new row key.
+        All complex expressions (expressions that aren't simply row fields) must
+        be passed in as a keyword argument and are assigned to a field.
 
         Parameters
         ----------
-        key_expr : str or :class:`.Expression`
-            Field name or expression to use as new row key.
+        exprs : args of :obj:`str` or :class:`.Expression`
+            Row fields to group on.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: s/on/by, here and below

File: python/hail/matrixtable.py [hunk 9]
```diff
@@ -1735,15 +1748,19 @@ def group_cols_by(self, *exprs, **named_exprs):
...
+                raise ExpressionException("method 'group_cols_by' expects keyword arguments for complex expressions")
             key = ast[0].name if isinstance(ast[0], Reference) else ast[0].selection
-            groups.append((key, e))
-        for k, e in named_exprs.items():
-            e = to_expr(e)
-            analyze('MatrixTable.group_cols_by', e, self._col_indices)
-            groups.append((k, e))
+            if key in new_keys:
+                raise ExpressionException("method 'group_cols_by' found duplicate field: {}".format(key))
+            new_keys.append(key)
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: same comments as group_rows_by

File: python/hail/matrixtable.py [hunk 10]
```diff
@@ -147,22 +148,21 @@ def aggregate(self, **named_exprs):
...
             analyze('GroupedMatrixTable.aggregate', v, self._grouped_indices,
                     {self._parent._row_axis, self._parent._col_axis})
             strs.append('{} = {}'.format(escape_id(k), v._ast.to_hql()))
 
-        key_strs = ['{} = {}'.format(escape_id(id), e._ast.to_hql()) for id, e in self._groups]
         if self._grouped_indices == self._parent._row_indices:
             # group rows
             return cleanup(
-                MatrixTable(base._jvds.groupVariantsBy(','.join(key_strs), ',\n'.join(strs), self._partition_key)))
+                MatrixTable(base.key_rows_by(*self._keys)._jvds.groupVariantsBy(',\n'.join(strs))))
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: what do we do about the partition key?

File: python/hail/matrixtable.py [hunk 11]
```diff
@@ -143,26 +335,43 @@ def aggregate(self, **named_exprs):
...
+        if self._col_keys is not None:
+            fixed_fields.extend(self._col_keys.keys())
+        else:
+            fixed_fields.extend([f.name for f in self._parent.col_schema.fields])
+
+        base, _ = self._parent._process_joins(*named_exprs.values())
         for k, v in named_exprs.items():
-            analyze('GroupedMatrixTable.aggregate', v, self._grouped_indices,
+            if k in fixed_fields:
+                raise ExpressionException("GroupedMatrixTable.group_cols_by cannot assign duplicate field '{}'".format(k))
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: wrong method name

File: python/hail/matrixtable.py [hunk 12]
```diff
@@ -143,26 +335,43 @@ def aggregate(self, **named_exprs):
...
+            analyze('GroupedMatrixTable.aggregate', v, self._fixed_indices(),
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: I think the _fixed_indices is wrong here. The aggregation expressions should always be entry-indexed.

File: python/hail/matrixtable.py [hunk 13]
```diff
@@ -143,26 +335,43 @@ def aggregate(self, **named_exprs):
...
-            return cleanup(
-                MatrixTable(base._jvds.groupVariantsBy(','.join(key_strs), ',\n'.join(strs), self._partition_key)))
-        else:
-            assert self._grouped_indices == self._parent._col_indices
-            # group cols
-            return cleanup(MatrixTable(base._jvds.groupSamplesBy(','.join(key_strs), ',\n'.join(strs))))
+        base, rename = self._process_keys(base)
+
+
+        if self._col_keys is not None:
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: this should be if/else if the two are exclusive for now.

File: python/hail/matrixtable.py [hunk 14]
```diff
@@ -25,33 +25,233 @@ class GroupedMatrixTable(object):
...
+            return mt.rename(renamed)
+        return left, cleanup
+
+    def describe(self):
+        """Print information about grouped matrix table."""
+
+        if self._row_keys is None:
+            rowstr = "NA"
+        else:
+            rowstr = "\n".join(["{}: {}".format(k, v) for k, v in self._row_keys.items()])
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/matrixtable.py]: What happens if you group by a nested field and describe?

File: python/hail/matrixtable.py (hunk 12 above)
Comment (catoverdrive) [reply python/hail/matrixtable.py]: those are the non-aggregation entries, as we talked about. So in the group_rows_by case they'd be the column indices, which are valid expressions to use outside of an aggregator.

Comment (catoverdrive) [reply python/hail/matrixtable.py]: *non-aggregation axes

Comment (tpoterba) [reply python/hail/matrixtable.py]: ahhh, right. Nevermind.

File: python/hail/matrixtable.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (catoverdrive) [reply python/hail/matrixtable.py]: I think I'm just going to make it show the new fields names and types for now. Not sure how to print the expr in a nice way here.

File: python/hail/matrixtable.py [hunk 15]
```diff
@@ -25,33 +25,233 @@ class GroupedMatrixTable(object):
...
-    def __init__(self, parent, groups, grouped_indices):
+    def __init__(self, parent, row_keys=None, col_keys=None):
         self._parent = parent
-        self._groups = groups
-        self._grouped_indices = grouped_indices
+        self._row_keys = row_keys
+        self._col_keys = col_keys
         self._partitions = None
         self._partition_key = None
-        self._fields = {}
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply python/hail/matrixtable.py]: The interface hasn't really changed? It's still invoked by MatrixTable.group_rows_by, and you had a `.partition_key()` method to change the default on GroupedMatrixTable which still seems like the right thing. I'm not really sure that _partitions is used. key_rows_by doesn't currently let you specify how many partitions you want. I can take it out for now.

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/matrixtable.py]: We've changed the semantics of the interface, I think -- the previous interface didn't allow you to do anything but partition_hint / partiton_by / aggregate. Now we've made it possible to group both rows and cols (though haven't written down implementations yet).

If somebody tries to group_rows and then group cols, it'll lose the partitioning. That's what I'm worried about

[the remark this answers is not in the document]
Comment (catoverdrive) [reply python/hail/matrixtable.py]: oh, good point. I'll add.

File: python/hail/matrixtable.py [hunk 16]
```diff
@@ -24,41 +24,232 @@ class GroupedMatrixTable(object):
...
+            self._partition_key)
+
+        print(s)
+        self._parent.describe()
+
+
+    @typecheck_method(exprs=oneof(str, Expression),
+                      named_exprs=oneof(str, Expression))
+    def group_rows_by(self, *exprs, **named_exprs):
+        """Group rows, used with :meth:`.GroupedMatrixTable.aggregate`
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: needs a period at the end.

File: python/hail/matrixtable.py [hunk 17]
```diff
@@ -24,41 +24,232 @@ class GroupedMatrixTable(object):
...
 
     def partition_by(self, *fields):
         """Set the partition key.
 
         Parameters
         ----------
         fields : varargs of :obj:`str`
-            Fields to partition by. Must be a prefix of the key.
+            Row partition key. Must be a a prefix of the key. By default, the
+            entire key is the partition key.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: flip the order -- the partition key is the entire key.

File: python/hail/matrixtable.py [hunk 18]
```diff
@@ -526,34 +727,48 @@ def key_cols_by(self, *keys):
...
+    @typecheck_method(keys=oneof(str, Expression),
+                      partition_key=nullable(oneof(oneof(str, Expression), listof(oneof(str, Expression)))))
+    def key_rows_by(self, *keys, partition_key=None):
         """Key rows by a new set of fields.
 
         Parameters
         ----------
         keys : varargs of :obj:`str`
             Row fields to key by.
+        partition_key : :obj:`str` or :obj:`list` of :obj:`str`, optional
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: docs are wrong, takes expressions too. A bit annoying to write that out. Can you also fix the docs for keys above?

File: python/hail/tests.py [hunk 19]
```diff
@@ -688,6 +688,82 @@ def test_range(self):
...
+    def test_named_fields_work_correctly(self):
+        mt = self.get_groupable_matrix()
+        a = mt.group_rows_by(group5=(mt['group2']['a'] + 1)).aggregate(c=hl.agg.sum(mt['c']))
+        self.assertEqual(a.count_rows(), 6)
+        self.assertEqual(a.row_key[0], 'group5')
+
+        b = mt.group_cols_by(group5=(mt['group4']['a'] + 1)).aggregate(c=hl.agg.sum(mt['c']))
+        self.assertEqual(b.count_cols(), 6)
+        self.assertEqual(b.col_key[0], 'group5')
+
```
Comment (tpoterba) [inline python/hail/tests.py]: nice tests.

File: src/main/scala/is/hail/variant/MatrixTable.scala [hunk 20]
```diff
@@ -661,65 +661,63 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
...
-    // FIXME: this shuffles twice
     val newRVType = newMatrixType.rvRowType
-    val rdd = keyedRDD
-      .aggregateByKey(zero)(seqOp, combOp)
-      .mapPartitions { it =>
+    val localRVType = rvRowType
+    val selectIdx = matrixType.orvdType.kRowFieldIdx
+    val keyOrd = matrixType.orvdType.kRowOrd
+
+    val newRVD = rvd.mapPartitionsPreservesPartitioning(newMatrixType.orvdType) { it =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/MatrixTable.scala]: a note that the old model benefited from map-side combine resulting in a smaller shuffle. We'll optimize this later.

Comment (tpoterba) [reply src/main/scala/is/hail/variant/MatrixTable.scala]: (no action necessary)

**Comment by tpoterba** [comment]
got some rebasing to do

**Pull request merged**